hbase-server/src/test/java/org/apache/hadoop/hbase/master/TestMasterShutdown.java

```diff
@@ -151,19 +152,46 @@ public void testMasterShutdownBeforeStartingAnyRegionServer() throws Exception {
-      // the server manager in HMaster is not initialized by the time shutdown() RPC (below) is
-      // made to the master. The suspected reason as to why it was uncommon before HBASE-18095
-      // is because the connection creation with ZK registry is so slow that by then the server
-      // manager is usually init'ed in time for the RPC to be made. For now, adding an explicit
-      // wait() in the test, waiting for the server manager to become available.
-      final long timeout = TimeUnit.MINUTES.toMillis(10);
-      assertNotEquals("Timeout waiting for server manager to become available.",
-        -1, Waiter.waitFor(htu.getConfiguration(), timeout,
-          () -> masterThread.getMaster().getServerManager() != null));
-      htu.getConnection().getAdmin().shutdown();
+      final CompletableFuture<Void> shutdownFuture = CompletableFuture.runAsync(() -> {
+        // Switching to master registry exacerbated a race in the master bootstrap that can result
+        // in a lost shutdown command (HBASE-8422, HBASE-23836). The race is essentially because
+        // the server manager in HMaster is not initialized by the time shutdown() RPC (below) is
+        // made to the master. The suspected reason as to why it was uncommon before HBASE-18095
+        // is because the connection creation with ZK registry is so slow that by then the server
+        // manager is usually init'ed in time for the RPC to be made. For now, adding an explicit
+        // wait() in the test, waiting for the server manager to become available.
+        final long timeout = TimeUnit.MINUTES.toMillis(10);
+        assertNotEquals("timeout waiting for server manager to become available.", -1,
+          htu.waitFor(timeout, () -> masterThread.getMaster().getServerManager() != null));
+
+        // Master has come up far enough that we can terminate it without creating a zombie.
+        LOG.debug("Attempting to establish connection.");
+        try {
+          // HBASE-24327 : (Resolve Flaky connection issues)
+          // shutdown() RPC can have flaky ZK connection issues.
+          // e.g
+          // ERROR [RpcServer.priority.RWQ.Fifo.read.handler=1,queue=1,port=53033]
+          // master.HMaster(2878): ZooKeeper exception trying to set cluster as down in ZK
+          // org.apache.zookeeper.KeeperException$SystemErrorException:
+          // KeeperErrorCode = SystemError
+          //
+          // However, even when above flakes happen, shutdown call does get completed even if
+          // RPC call has failure. Hence, subsequent retries will never succeed as HMaster is
+          // already shutdown. Hence, it can fail. To resolve it, after making one shutdown()
+          // call, we are ignoring IOException.
+          htu.getConnection().getAdmin().shutdown();
+          LOG.info("Shutdown RPC sent.");
+        } catch (IOException | CompletionException e) {
+          LOG.warn("Failed to establish connection.", e);
```

Comment:
Your above comment says "make one shutdown and ignore failures" but the warning here makes the reader think the shutdown call failed (or, that something failed. the message doesn't say what was attempting to establish a connection, or to what).

The `IOException` is highly generic -- we don't actually know if it was a connection failure, a failure on the server side processing the request, or a failed on the client to interpret the result.

Reply:
@ndimiduk Are you in favor of checking the root cause like the first commit: https://github.com/apache/hbase/pull/1690/commits/04260b5bf95670e9493400214c9b682267df8de2#diff-144b3c3611227c12067b788b985183f9R182  &&  https://github.com/apache/hbase/commit/04260b5bf95670e9493400214c9b682267df8de2#diff-144b3c3611227c12067b788b985183f9R207 ?
cause message `Failed contacting masters after 1 attempts` with MasterRegistry being wrapped inside RetriesExhaustedException.

Reply:
Thanks for the links to the previous commits. GH makes it difficult for me to follow previous conversation.

Okay, so if you move this code out of the `CompletableFuture`, the `CompletionException` can just be re-thrown. It indicates a client-side problem, and there's nothing the test should try to recover.

I would not catch `IOException` either, just let it bubble up. Instead, I would focus on meaningful subclasses of `IOException`. `RetriesExhaustedException` is a good place to work from: you know your client made some effort. There's potentially multiple cause instances under there, so i guess just pick one to work with, probably the last one. If it's a descendent of `java.net.SocketException`, you know the client couldn't get anywhere -- how should the test behave if the the master is not running?

After that, I'm not really sure what's thrown in what cases. Our API's checked exception definitions aren't strong enough for me know by reading the interfaces. However, pretty much anything else means the client managed to get the RPC over to the server. I think that fact alone is enough to consider this part of the test has succeeded at its goal. Until the myriad other issues in master startup and shutdown are resolved, I think this is the best the client can hope for (for what it's worth, I think this test will continue to be flakey until those master-side problems are solved, and they cannot be resolved perfectly by client-side gymnastics).

You've looked at it, and seen the errors and test failures, more recently than I have, so what do you think? What brought you to this ticket in the first place? Are there more specific subclasses of `IOException` that are thrown, which you can use to reasonably address that specific condition? Paste the stack traces into Jira and the commit message so we can follow your effort.

And thank you for your effort -- test fixing is usually thankless drudgery but it makes all of our lives better :)

Reply:
Thanks @ndimiduk 
Let me put the code reference of the link that I mentioned. This is what I had in my first commit:

```
          } catch (IOException|CompletionException e) {
            LOG.error("Failed to establish connection.");
            if (connectionFailedWithMaster(e)) {
              return true;
            }
          }
```

```
  private boolean connectionFailedWithMaster(Exception e) {
    if (e instanceof RetriesExhaustedException) {
      Throwable cause = e.getCause();
      if (cause instanceof MasterRegistryFetchException) {
        cause = cause.getCause();
        if (cause instanceof RetriesExhaustedException) {
          final String message = cause.getMessage();
          return message != null && message
            .startsWith("Failed contacting masters after 1 attempts");
        }
      }
    }
    return false;
  }
```

Reply:
@ndimiduk @bharathv 
I will remove `CompletionException` from catch list. And what about IOException? Good to follow above code to find out the root cause message and accordingly handle it?
This is what I had in my first commit and then Bharath and I came to conclusion that it's bit hacky and why not just handle IOException and not worry about cause. But now that I again think of it, you are right, although there are less chances of having different root cause for IOException but still it's good to be specific, that way not too hacky right?

Reply:
I think the MasterRegistryFetchException is only valid if we re-try shutdown() rpcs. First shutdown() throws an IOException (some subclass of it) and shuts down the master, the second (and later) shutdowns can't find a master to issue a shutdown and the stub fails and throws this MasterRegistryFetchException. 

Now that we decided to only do a single shutdown, we should either see the specific exception (derivate of IOException when shutdown failed) or no exception at all (if shutdown is a success, which is 99% of the test runs). I think we are good if we catch and log the first exception (don't remember the exact checked exception type).

The reason I suggested to keep the checks simple and catch the exception and log is because, if there was any issue in the RPC, that'd anyway reflect in the failure of shutdown join thread and the test fails and we have the exception logged (we don't need to go fancy on checking exception.getCause() recursively).

I think what Nick is suggesting is to catch the specific subclass of IOException that is thrown if shutdown() fails.  @ndimiduk Correct me if I got you wrong..

Reply:
Sounds good, but we can't catch the root cause Exception directly, which in our case is ConnectionClosedException. So we catch RetriesExhaustedException first and look for root cause:

```
      } catch (RetriesExhaustedException e) {
        if (e.getCause() instanceof ConnectionClosedException) {
          LOG.info("Connection is Closed, ZK is not reachable.", e);
        }
      }
```